File: backend/src/models/Place.model.ts
import mongoose, { Schema, Document, Types, model } from "mongoose";
import { Visibility } from "../enums";
import { IPlace } from '../interfaces';

export type IPlaceDocument = IPlace & Document;

export const PlaceSchema = new Schema({

    name: {
        type: String,
        required: true
    },
    userId: {
        type: String,
        required: true
    },
    withWho: {
        type: String,
        required: false,
    },
    photo: {
        type: String,
        required: false
    },
    coordinates: {
        longitude: {
            type: Number,
            required: true
        },
        latitude: {
            type: Number,
            required: true
        }
    },
    date: {
        type: Date,
        require: true
    },
    visibility: {
        type: String,
        enum: Object.values(Visibility),
        require: true
    },
    note: {
        type: String,
        required: false
    }
}, {
    versionKey: false,
    timestamps: true
});

export const PlaceModel = model<IPlaceDocument>('Place', PlaceSchema);

File: frontend/la_map/lib/models/place_model.dart
import 'package:google_maps_flutter/google_maps_flutter.dart';

class Place {
  final String id;
  final String name;
  final String userId;
  final String? withWho;
  final String? photo;
  final LatLng coordinates;
  final DateTime date;
  final String visibility;
  final String? note;

  const Place(
      {required this.id,
      required this.name,
      required this.userId,
      this.withWho,
      this.photo,
      required this.coordinates,
      required this.date,
      required this.visibility,
      this.note});

  factory Place.fromJson(Map<String, dynamic> json) {
    return Place(
      id: json['_id'],
      name: json['name'],
      userId: json['userId'],
      withWho: json['withWho'],
      photo: json['photo'],
      date: DateTime.parse(json['date']),
      note: json['note'],
      coordinates: LatLng(Coordinates.fromJson(json['coordinates']).latitude,
          Coordinates.fromJson(json['coordinates']).longitude),
      visibility: json['visibility'],
    );
  }
}

class Coordinates {
  final double longitude;
  final double latitude;

  Coordinates({required this.latitude, required this.longitude});

  factory Coordinates.fromJson(Map<String, dynamic> json) {
    return Coordinates(
        latitude: json['latitude'], longitude: json['longitude']);
  }
}
